pkg/genericapiserver/config_selfclient.go

```diff
@@ -75,24 +75,46 @@ func (s *SecureServingInfo) NewSelfClientConfig(token string) (*restclient.Confi
 		BearerToken: token,
 	}
 
-	// find certificate for host: either explicitly given, from the server cert bundle or one of the SNI certs
+	// find certificate for host: either explicitly given, from the server cert bundle or one of the SNI certs,
+	// but only return CA:TRUE certificates.
 	var derCA []byte
 	if s.CACert != nil {
 		derCA = s.CACert.Certificate[0]
 	}
 	if derCA == nil && s.Cert != nil {
-		x509Cert, err := x509.ParseCertificate(s.Cert.Certificate[0])
+		chain, err := parseChain(s.Cert.Certificate)
 		if err != nil {
 			return nil, fmt.Errorf("failed to parse server certificate: %v", err)
 		}
 
-		if (net.ParseIP(host) != nil && certMatchesIP(x509Cert, host)) || certMatchesName(x509Cert, host) {
-			derCA = s.Cert.Certificate[0]
+		if (net.ParseIP(host) != nil && certMatchesIP(chain[0], host)) || certMatchesName(chain[0], host) {
+			if trustedChain(chain) {
+				return clientConfig, nil
+			}
+
+			ca, err := findCA(chain)
+			if err != nil {
+				return nil, fmt.Errorf("no CA certificate found in server certificate bundle: %v", err)
+			}
+			derCA = ca.Raw
 		}
 	}
 	if derCA == nil && net.ParseIP(host) == nil {
```

Comment:
should do this first, since the SNI cert will be preferred if it matches the host